operator/pkg/ciliumendpointslice/endpointslice.go

```diff
@@ -48,263 +48,244 @@ const (
+		return
 	}
-	cesStore := ciliumEndpointSliceInit(clientset.CiliumV2alpha1(), ctx, wg)
-	ceSliceStore = cesStore
-
-	// List all existing CESs from the api-server and cache it locally.
-	// This sync should happen before starting CEP watcher, because CEP watcher
-	// emits the existing CEPs as newly added CEPs. If we don't have local sync
-	// cesManager would assume those are new CEPs and may create new CESs for those CEPs.
-	// This situation ends up having duplicate CEPs in different CESs. Hence, we need
-	// to sync existing CESs before starting a CEP watcher.
-	syncCESsInLocalCache(cesStore, manager)
-	return &CiliumEndpointSliceController{
-		clientV2:                 clientset.CiliumV2(),
-		clientV2a1:               clientset.CiliumV2alpha1(),
-		reconciler:               newReconciler(clientset.CiliumV2alpha1(), manager),
-		Manager:                  manager,
-		queue:                    rlQueue,
-		ciliumEndpointSliceStore: cesStore,
-		slicingMode:              slicingMode,
-		workerLoopPeriod:         1 * time.Second,
+	if c.endpointsMappingInitialized {
+		touchedCESs := c.manager.UpdateCEPMapping(k8s.ConvertCEPToCoreCEP(cep), cep.Namespace)
+		c.enqueueCESReconciliation(touchedCESs)
+	} else {
+		c.preInitEnqueuedEndpointsEvents = append(c.preInitEnqueuedEndpointsEvents, endpointEvent{event: updateEvent, cep: cep})
 	}
 }
 
-// start the worker thread, reconciles the modified CESs with api-server
-func (c *CiliumEndpointSliceController) Run(ces cache.Indexer, stopCh <-chan struct{}) {
-	log.Info("Bootstrap ces controller")
-	defer utilruntime.HandleCrash()
-	defer c.queue.ShutDown()
+func (c *Controller) onEndpointDelete(cep *cilium_api_v2.CiliumEndpoint) {
+	if c.endpointsMappingInitialized {
+		touchedCES := c.manager.RemoveCEPMapping(k8s.ConvertCEPToCoreCEP(cep), cep.Namespace)
+		c.enqueueCESReconciliation([]CESName{touchedCES})
+	} else {
+		c.preInitEnqueuedEndpointsEvents = append(c.preInitEnqueuedEndpointsEvents, endpointEvent{event: deleteEvent, cep: cep})
+	}
+}
 
-	// Cache CiliumEndpointStore Interface locally
-	c.ciliumEndpointStore = ces
+func (c *Controller) onSliceUpdate(ces *capi_v2a1.CiliumEndpointSlice) {
+	c.enqueueCESReconciliation([]CESName{NewCESName(ces.Name)})
+}
 
-	// On operator warm boot, remove stale CEP entries present in CES
-	c.removeStaleAndDuplicatedCEPEntries()
+func (c *Controller) onSliceDelete(ces *capi_v2a1.CiliumEndpointSlice) {
+	c.enqueueCESReconciliation([]CESName{NewCESName(ces.Name)})
+}
 
-	log.WithFields(logrus.Fields{
-		logfields.CESSliceMode: c.slicingMode,
-	}).Info("Starting CES controller reconciler.")
+func (c *Controller) enqueueCESReconciliation(cess []CESName) {
+	for _, ces := range cess {
+		c.logger.WithFields(logrus.Fields{
+			logfields.CESName: ces.string(),
+		}).Debug("Enqueueing CES (if not empty name)")
+		if ces.Name != "" {
+			if c.enqueuedAt[ces].IsZero() {
+				c.enqueuedAt[ces] = time.Now()
```

Comment:
> I don't see a case in which the result without synchronization cannot be achieved with synchronization. I'd rather not add a lock here just to protect queue delay metric.

The write operations themselves (`c.enqueuedAt[ces] = time.Now()` and `c.enqueuedAt[ces] = time.Time{}`) are not atomic. So we can end up with a value in `c.enqueuedAt[ces]` that is neither `time.Now()` nor `time.Time{}`.
That's because `time.Time` is defined as a struct:

```
type Time struct {
	wall uint64
	ext  int64

	loc *Location
}
```

And if that happens, `getAndReset` might return a totally bogus time interval when doing `time.Since(enqueued).Seconds()` the next time is called for the same CES.

Reply:
Added a lock